File: libopaecxx/src/shared_buffer.cpp
// Copyright(c) 2018, Intel Corporation
//
// Redistribution  and  use  in source  and  binary  forms,  with  or  without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of  source code  must retain the  above copyright notice,
//   this list of conditions and the following disclaimer.
// * Redistributions in binary form must reproduce the above copyright notice,
//   this list of conditions and the following disclaimer in the documentation
//   and/or other materials provided with the distribution.
// * Neither the name  of Intel Corporation  nor the names of its contributors
//   may be used to  endorse or promote  products derived  from this  software
//   without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.
#include <algorithm>
#include <cstring>

#include <opae/cxx/core/shared_buffer.h>
#include <exception>

namespace opae {
namespace fpga {
namespace types {

shared_buffer::~shared_buffer() { release(); }

shared_buffer::ptr_t shared_buffer::allocate(handle::ptr_t handle, size_t len,
                                             bool read_only) {
  ptr_t p;

  if (!handle) {
    throw std::invalid_argument("handle object is null");
  }

  if (!len) {
    throw except(OPAECXX_HERE);
  }

  uint8_t *virt = nullptr;
  uint64_t io_address = 0;
  uint64_t wsid = 0;

  int flags = 0;
  if (read_only) {
      flags |= FPGA_BUF_READ_ONLY;
  }

  fpga_result res = fpgaPrepareBuffer(
      handle->c_type(), len, reinterpret_cast<void **>(&virt), &wsid, flags);
  ASSERT_FPGA_OK(res);
  res = fpgaGetIOAddress(handle->c_type(), wsid, &io_address);
  ASSERT_FPGA_OK(res);
  p.reset(new shared_buffer(handle, len, virt, wsid, io_address));

  return p;
}

shared_buffer::ptr_t shared_buffer::attach(handle::ptr_t handle, uint8_t *base,
                                           size_t len,
                                           bool read_only) {
  ptr_t p;

  uint8_t *virt = base;
  uint64_t io_address = 0;
  uint64_t wsid = 0;

  int flags = FPGA_BUF_PREALLOCATED;
  if (read_only) {
      flags |= FPGA_BUF_READ_ONLY;
  }

  fpga_result res =
      fpgaPrepareBuffer(handle->c_type(), len, reinterpret_cast<void **>(&virt),
                        &wsid, flags);

  ASSERT_FPGA_OK(res);
  res = fpgaGetIOAddress(handle->c_type(), wsid, &io_address);
  ASSERT_FPGA_OK(res);
  p.reset(new shared_buffer(handle, len, virt, wsid, io_address));

  return p;
}

void shared_buffer::release() {
  // If the allocation was successful.
  if (virt_ && handle_) {
    auto res = fpgaReleaseBuffer(handle_->c_type(), wsid_);
    if (res == FPGA_OK) {
      virt_ = nullptr;
      len_ = 0;
      wsid_ = 0;
      io_address_ = 0;
    } else {
      std::cerr << "Error while calling fpgaReleaseBuffer: " << fpgaErrStr(res)
                << "\n";
    }
  }
}

void shared_buffer::fill(int c) { std::fill(virt_, virt_ + len_, c); }

int shared_buffer::compare(shared_buffer::ptr_t other, size_t len) const {
  return std::equal(virt_, virt_ + len, other->virt_) ? 0 : 1;
}

shared_buffer::shared_buffer(handle::ptr_t handle, size_t len, uint8_t *virt,
                             uint64_t wsid, uint64_t io_address)
    : handle_(handle),
      len_(len),
      virt_(virt),
      wsid_(wsid),
      io_address_(io_address) {}

}  // end of namespace types
}  // end of namespace fpga
}  // end of namespace opae

File: python/opae.http/opae/http/shared_buffer.py
#!/usr/bin/env python3
# Copyright(c) 2023, Intel Corporation
#
# Redistribution  and  use  in source  and  binary  forms,  with  or  without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of  source code  must retain the  above copyright notice,
#   this list of conditions and the following disclaimer.
# * Redistributions in binary form must reproduce the above copyright notice,
#   this list of conditions and the following disclaimer in the documentation
#   and/or other materials provided with the distribution.
# * Neither the name  of Intel Corporation  nor the names of its contributors
#   may be used to  endorse or promote  products derived  from this  software
#   without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
# IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
# LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
# CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
# SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
# INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
# CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import base64
import requests

from opae.http.conversion import to_json_obj
from opae.http.fpga_remote_id import fpga_remote_id
import opae.http.constants as constants


class shared_buffer():
    def __init__(self, h, buf_id, length, debug):
        self.handle = h
        self.buf_id = buf_id
        self.length = length
        self.debug = debug

    def destroy(self):
        shared = self.handle.__dict__['_shared_buffers']
        if self not in shared:
            msg = f'buffer {self} is not tracked by {self.handle}'
            constants.raise_for_error(constants.FPGA_INVALID_PARAM, msg)

        tok = self.handle.__dict__['_token']
        url = tok.__dict__['_url'] + '/fpga/v1/handle/buffers/release'
        if self.debug:
            print(url)

        req = {'handle_id': self.handle.handle_id.to_json_obj(),
               'buf_id': self.buf_id.to_json_obj()}

        resp = requests.post(url, json=req)

        resp.raise_for_status()

        jobj = resp.json()

        constants.raise_for_error(jobj['result'], 'fpgaReleaseBuffer returned')
        shared.remove(self)

    def ioaddr(self):
        tok = self.handle.__dict__['_token']
        url = tok.__dict__['_url'] + '/fpga/v1/handle/buffers/ioaddress'
        if self.debug:
            print(url)

        req = {'handle_id': self.handle.handle_id.to_json_obj(),
               'buf_id': self.buf_id.to_json_obj()}

        resp = requests.post(url, json=req)

        resp.raise_for_status()

        jobj = resp.json()

        constants.raise_for_error(jobj['result'], 'fpgaGetIOAddress returned')
        return int(jobj['ioaddr'])

    def memset(self, offset, c, n):
        tok = self.handle.__dict__['_token']
        url = tok.__dict__['_url'] + '/fpga/v1/handle/buffers/memset'
        if self.debug:
            print(url)

        req = {'handle_id': self.handle.handle_id.to_json_obj(),
               'buf_id': self.buf_id.to_json_obj(),
               'offset': str(offset),
               'c': c,
               'n': str(n)}

        resp = requests.post(url, json=req)

        resp.raise_for_status()

        jobj = resp.json()

        constants.raise_for_error(jobj['result'], 'fpgaBufMemSet returned')

    def memcpy(self, dest_offset, src_data, num_bytes):
        tok = self.handle.__dict__['_token']
        url = tok.__dict__['_url'] + '/fpga/v1/handle/buffers/memcpy'
        if self.debug:
            print(url)

        encoded = base64.b64encode(src_data)

        req = {'handle_id': self.handle.handle_id.to_json_obj(),
               'dest_buf_id': self.buf_id.to_json_obj(),
               'dest_offset': str(dest_offset),
               'src': encoded.decode(),
               'n': str(num_bytes)}

        resp = requests.post(url, json=req)

        resp.raise_for_status()

        jobj = resp.json()

        constants.raise_for_error(jobj['result'], 'fpgaBufMemCpyToRemote returned')

    def poll(self, offset, width, mask, expected_value, sleep_interval, loops_timeout):
        tok = self.handle.__dict__['_token']
        url = tok.__dict__['_url'] + '/fpga/v1/handle/buffers/poll'
        if self.debug:
            print(url)

        req = {'handle_id': self.handle.handle_id.to_json_obj(),
               'buf_id': self.buf_id.to_json_obj(),
               'offset': str(offset),
               'width': width,
               'mask': str(mask),
               'expected_value': str(expected_value),
               'sleep_interval': str(sleep_interval),
               'loops_timeout': str(loops_timeout)}

        resp = requests.post(url, json=req)

        resp.raise_for_status()

        jobj = resp.json()

        result = constants.fpga_result_from_str(jobj['result'])

        if result not in [constants.FPGA_OK, constants.FPGA_NOT_FOUND]:
            constants.raise_for_error(result, 'fpgaBufPoll returned')

        return result

    def memcmp(self, other, bufa_offset, bufb_offset, n):
        tok = self.handle.__dict__['_token']
        url = tok.__dict__['_url'] + '/fpga/v1/handle/buffers/compare'
        if self.debug:
            print(url)

        req = {'handle_id': self.handle.handle_id.to_json_obj(),
               'bufa_id': self.buf_id.to_json_obj(),
               'bufa_offset': str(bufa_offset),
               'bufb_id': other.buf_id.to_json_obj(),
               'bufb_offset': str(bufb_offset),
               'n': str(n)}

        resp = requests.post(url, json=req)

        resp.raise_for_status()

        jobj = resp.json()

        constants.raise_for_error(jobj['result'], 'fpgaBufMemCmp returned')

        return jobj['cmpResult']

    def write_pattern(self, pattern_name):
        tok = self.handle.__dict__['_token']
        url = tok.__dict__['_url'] + '/fpga/v1/handle/buffers/pattern/write'
        if self.debug:
            print(url)

        req = {'handle_id': self.handle.handle_id.to_json_obj(),
               'buf_id': self.buf_id.to_json_obj(),
               'pattern_name': pattern_name}

        resp = requests.post(url, json=req)

        resp.raise_for_status()

        jobj = resp.json()

        constants.raise_for_error(jobj['result'], 'fpgaBufWritePattern returned')
